app/knight.py

```diff
@@ -0,0 +1,42 @@
+from typing import List, Dict
+
+
+class Knight:
+    def __init__(
+            self,
+            name: str,
+            hp: int,
+            power: int,
+            armour: List[Dict[str, int]],
+            weapon: Dict[str, int],
+            potion: Dict[str, Dict[str, int]] = None
+    ) -> None:
+        self.name = name
+        self.hp = hp
+        self.power = power
+        self.armour = armour
+        self.weapon = weapon
+        self.potion = potion
+        self.protection = 0
+
+    def apply_armour(self) -> None:
+        self.protection = sum(piece["protection"] for piece in self.armour)
+
+    def apply_weapon(self) -> None:
+        self.power += self.weapon["power"]
+
+    def apply_potion(self) -> None:
+        if self.potion:
+            potion = self.potion.get("effect", {})
+            for effect, value in potion.items():
+                if effect == "power":
+                    self.power += value
+                elif effect == "protection":
+                    self.protection += value
+                elif effect == "hp":
+                    self.hp += value
```

Comment:
Its ok, but I dont like how you are increasing levels.
1st thing that you can do is simply do it reversed.
```suggestion
        if not self.potion:
            return

        potion = self.potion.get("effect", {})
        for effect, value in potion.items():
            if effect == "power":
                self.power += value
            elif effect == "protection":
                self.protection += value
            elif effect == "hp":
                self.hp += value
```
And if we think about optimizing I would use mapping to simplify adding new effects
```suggestion
def apply_potion(self) -> None:
    if not self.potion:
        return
    
    effect_map = {
        "power": "power",
        "protection": "protection",
        "hp": "hp"
    }

    for effect, value in self.potion.get("effect", {}).items():
        if effect in effect_map:
            setattr(self, effect_map[effect],
                    getattr(self, effect_map[effect]) + value)
```